webview-ui/connectionsetup.js

```diff
@@ -29,17 +29,40 @@ function init() {
 }
 
 function onChangeConnectionId() {
+  // If connection ID is manually changed, we should stop updating it when the URL changes
+  byId('shouldGenerateConnectionId').value = 'false';
   saveState();
+  toggleSaveConnectionButton();
 }
 
 function onChangeServerUrl() {
   saveState();
-  byId('connectionId').value = sanitize(byId('serverUrl').value);
+  if (byId('shouldGenerateConnectionId').value === 'true') {
+    byId('connectionId').value = sanitize(byId('serverUrl').value);
+  }
   toggleGenerateTokenButton();
+  toggleSaveConnectionButton();
 }
 
 function toggleGenerateTokenButton() {
-  byId('generateToken').disabled = !byId('serverUrl').validity.valid;
+  byId('generateToken').disabled = !hasValidServerUrl();
+}
+
+function hasValidServerUrl() {
+  /**
+   * @type {HTMLInputElement}
+   */
+  const serverUrlInput = byId('serverUrl');
+  return serverUrlInput.validity.valid && isValidUrl(serverUrlInput.value);
+}
+
+function isValidUrl(value) {
```

Comment:
I think it would be great to have tests for this method

Reply:
This might prove tricky: this file is only used in the webview and is completely separated from the TS module system.

Reply:
ack